### ISSUE TITLE
How to improve our loss function

### ISSUE THREAD

**Issue by c100rczyk** [c100rczyk/Team_project_PWR#26]
https://arxiv.org/abs/1503.03832

**Comment by c100rczyk** [comment]
Important information from the work:
1.  It is crucial to select hard triplets, that are active and can therefore contribute to improving the model.

**Comment by c100rczyk** [comment]
2. Use hard positive and hard negative images in triplets. 
Hard positive argmax_xip||f(xia) - f(xip)||^2 
Hard negative argmin_xin ||f(xia) - f(xin)||^2
Two possible choices:

+ Generate triplets offline every n steps, using the most
recent network checkpoint and computing the argmin
and argmax on a subset of the data
+ Generate triplets online. This can be done by select-
ing the hard positive/negative exemplars from within a
mini-batch.  (probably this second way is good for us).

**Comment by c100rczyk** [comment]
Important to note: 
Hard negatives examples can lead to bad local minima of model. Model can collapsed. 
SO: it is important (especially on early phase of training) to choose such negatives that:
distance_between(anchor, positive) < distance_between(anchor, negative)

**Comment by c100rczyk** [comment]
Proposition: 
In each batch find some hard positives examples (for example):
+ for each 32 mini-batch , get 10 examples of (most hard positives).

And get most hard negatives to the triplets that meet the condition : 
+ distance_between(anchor, positive) < distance_between(anchor, negative)

Train only on that examples in each batch

**Comment by c100rczyk** [comment]
Based on this we have maybe the answer (why sometimes it happend that our model did not learn). The accuracy top1 was 50%. So at the beginning model could get some hard negatives which had a strong negative impact on further learning and the model became blocked.

**Comment by c100rczyk** [comment]
Articles that i think will help with implementation: 
https://omoindrot.github.io/triplet-loss
https://arxiv.org/abs/1703.07737
And repository:
https://github.com/omoindrot/tensorflow-triplet-loss
